internal/config/etcd_source.go

```diff
@@ -0,0 +1,161 @@
+// limitations under the License.
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	ModeWatch = iota
+	ModeInterval
+)
+
+type EtcdSource struct {
+	etcdCli *clientv3.Client
+	ctx     context.Context
+	sync.RWMutex
+	currentConfig map[string]string
+
+	keyPrefix       string
+	refreshMode     int
+	refreshInterval time.Duration
+	eh              EventHandler
+}
+
+func NewEtcdSource(remoteInfo *EtcdInfo) (*EtcdSource, error) {
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:   remoteInfo.Endpoints,
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &EtcdSource{
+		etcdCli:         etcdCli,
+		ctx:             context.Background(),
+		keyPrefix:       remoteInfo.KeyPrefix,
+		refreshMode:     remoteInfo.RefreshMode,
+		refreshInterval: remoteInfo.RefreshInterval,
+	}, nil
+}
+
+// Delete implements ConfigSource
+func (es *EtcdSource) Delete(key string) error {
+	return nil
+}
+
+// GetConfigurationByKey implements ConfigSource
+func (es *EtcdSource) GetConfigurationByKey(key string) (string, error) {
+	es.RLock()
+	v, ok := es.currentConfig[key]
+	es.RUnlock()
+	if !ok {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return v, nil
+}
+
+// GetConfigurations implements ConfigSource
+func (es *EtcdSource) GetConfigurations() (map[string]string, error) {
+	configMap := make(map[string]string)
+	err := es.refreshConfigurations()
+	if err != nil {
+		return nil, err
+	}
+	if es.refreshMode == ModeInterval {
+		go es.refreshConfigurationsPeriodically()
+	}
+
+	es.RLock()
+	for key, value := range es.currentConfig {
+		configMap[key] = value
+	}
+	es.RUnlock()
+
+	return configMap, nil
+}
+
+// GetPriority implements ConfigSource
+func (es *EtcdSource) GetPriority() int {
+	return 1
+}
+
+// GetSourceName implements ConfigSource
+func (es *EtcdSource) GetSourceName() string {
+	return "RemoteSource"
+}
+
+// Set implements ConfigSource
+func (es *EtcdSource) Set(key string, value string) error {
+	return nil
+}
+
+func (es *EtcdSource) refreshConfigurations() error {
+	prefix := es.keyPrefix + "/config"
+	response, err := es.etcdCli.Get(es.ctx, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return err
+	}
+	newConfig := make(map[string]string, len(response.Kvs))
+	for _, kv := range response.Kvs {
+		key := string(kv.Key)
+		key = strings.TrimPrefix(key, prefix+"/")
+		newConfig[key] = string(kv.Value)
+		newConfig[formatKey(key)] = string(kv.Value)
+	}
+	return es.updateConfigurationAndFireEvent(newConfig)
+}
+
+func (es *EtcdSource) refreshConfigurationsPeriodically() {
+	ticker := time.Tick(es.refreshInterval)
+	log.Info("start refreshing configurations")
+	for range ticker {
+		err := es.refreshConfigurations()
+		if err != nil {
+			log.Error("can not pull configs", zap.Error(err))
+		}
+	}
+	log.Info("stop refreshing configurations")
```

Comment:
Looks like this goroutine can never quit until process exit?

Reply:
Maybe I should add the `Close` method for the `manager` to interrupt it.